### PR TITLE
Add Sorted Empirical CDF CRPS algorithm

### LINKED ISSUE #455: 🚀[FEA]: sort-based CRPS
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

Medium

### Please provide a clear description of problem you would like to solve.

Can we add a sort-based implementation of exact CRPS to the metrics? In theory this should give the same answer as the kernel method, but in O(n log n) time and memory rather than O(n^2).

earth2mip has an implementation here: https://github.com/NVIDIA/earth2mip/blob/main/earth2mip/crps.py. There are also some tests. I've used this for large-scale scoring. Would a PR be welcome?

### Describe any alternatives you have considered

_No response_

### PR DESCRIPTION
**Pull request by dallasfoster**
# Modulus Pull Request

## Description
This PR adds another method for computing the Continuous Ranked Probability Score. It lifts the implementation from Earth2MIP, https://github.com/NVIDIA/earth2mip/blob/main/earth2mip/crps.py, at the request from @nbren12.

Closes #455 
## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: modulus/metrics/general/crps.py [hunk 1]
```diff
@@ -124,6 +134,82 @@ def _crps_gaussian(mean: Tensor, std: Tensor, obs: Union[Tensor, np.ndarray]) ->
     return std * (2 * phi + d * Phi - 1.0 / torch.sqrt(torch.as_tensor(torch.pi)))
 
 
+@torch.jit.script
+def _crps_from_empirical_cdf(
+    pred: torch.Tensor, obs: torch.Tensor, dim: int = 0
+) -> torch.Tensor:
+    """Compute the exact CRPS using the CDF method
+
+    Uses this formula
```
Comment (nbren12) [inline modulus/metrics/general/crps.py:143]: Maybe worth moving some of the docstring into the public function

File: modulus/metrics/general/crps.py [hunk 2]
```diff
@@ -124,6 +134,82 @@ def _crps_gaussian(mean: Tensor, std: Tensor, obs: Union[Tensor, np.ndarray]) ->
     return std * (2 * phi + d * Phi - 1.0 / torch.sqrt(torch.as_tensor(torch.pi)))
 
 
+@torch.jit.script
+def _crps_from_empirical_cdf(
```
Comment (nbren12) [inline modulus/metrics/general/crps.py:138]: Looks the same as e2mip, but you added support for arbitrary dim. Make sense to me.

File: modulus/metrics/general/crps.py [hunk 3]
```diff
@@ -50,14 +50,24 @@ def kcrps(pred: Tensor, obs: Tensor, dim: int = 0):
...
-    n = pred.shape[0]
+    n = pred.shape[dim]
+    device = pred.device
     _crps = 0.0 * obs
     for i in range(n):
-        x_i = pred[i]
-        x_j = pred[i:]
-        _crps += torch.abs(x_i - obs) / n
-        _crps -= torch.sum(torch.abs(x_i[None] - x_j) / n, dim=0) / n
+        # x_i = pred[i]
```
Comment (nbren12) [inline modulus/metrics/general/crps.py:57]: I didn't review this closely. I suppose again just adding `dim` support?

**Review by nbren12 — APPROVED** [review]
LGTM. Thanks for adding!

File: modulus/metrics/general/crps.py (hunk 3 above)
Comment (dallasfoster) [reply modulus/metrics/general/crps.py:57]: Yes, this is just adding `dim` support.

File: modulus/metrics/general/crps.py [hunk 4]
```diff
@@ -291,15 +377,17 @@ def crps(
     Tensor
         Map of CRPS
     """
-    if method not in ["kernel", "histogram"]:
+    if method not in ["kernel", "sort", "histogram"]:
         raise ValueError("Method must either be 'kernel' or 'histogram'.")
```
Comment (NickGeneva) [inline modulus/metrics/general/crps.py:381]: or 'sort'

File: modulus/metrics/general/crps.py [hunk 5]
```diff
@@ -50,14 +50,24 @@ def kcrps(pred: Tensor, obs: Tensor, dim: int = 0):
...
-        x_i = pred[i]
-        x_j = pred[i:]
-        _crps += torch.abs(x_i - obs) / n
-        _crps -= torch.sum(torch.abs(x_i[None] - x_j) / n, dim=0) / n
+        # x_i = pred[i]
+        x_i = torch.index_select(
+            pred, dim, torch.tensor([i], device=device, dtype=torch.int32)
+        )
+
+        # x_j = pred[i:]
```
Comment (NickGeneva) [inline modulus/metrics/general/crps.py:62]: comment still needed?

**Review by NickGeneva — APPROVED** [review]
